usr/src/uts/common/fs/zfs/dsl_dataset.c

```diff
@@ -2501,11 +2501,38 @@ dsl_dataset_rollback_check(void *arg, dmu_tx_t *tx)
 	 * the latest snapshot is it.
 	 */
 	if (ddra->ddra_tosnap != NULL) {
-		char namebuf[ZFS_MAX_DATASET_NAME_LEN];
+		dsl_dataset_t *snapds;
 
-		dsl_dataset_name(ds->ds_prev, namebuf);
-		if (strcmp(namebuf, ddra->ddra_tosnap) != 0)
-			return (SET_ERROR(EXDEV));
+		/* Check if the target snapshot exists at all. */
+		error = dsl_dataset_hold(dp, ddra->ddra_tosnap, FTAG, &snapds);
+		if (error != 0) {
+			/* Distinguish between missing dataset and snapshot. */
+			if (error == ENOENT || error == EXDEV)
+				error = SET_ERROR(ESRCH);
+			dsl_dataset_rele(ds, FTAG);
+			return (error);
+		}
+		ASSERT(snapds->ds_is_snapshot);
+
+		/* Check if the snapshot is the latest snapshot indeed. */
+		if (snapds != ds->ds_prev) {
+			/*
+			 * See if the snapshot is a snapshot of the filesystem
+			 * or the snapshot is an origin of the filesystem.
```

Comment:
nit: `is *the* origin`

you might say something like, `distinguish between the case where the only problem is intervening snapshots (EEXIST) vs the snapshot can't possibly be a valid target for rollback (or doesn't exist) (ESRCH).`

Reply:
Thank you for the suggestion!